File: scripts/tools/check_glossary_duplicates.mjs
#!/usr/bin/env node
/**
 * check_glossary_duplicates.mjs — detect duplicate entries in docs/GLOSSARY.md.
 *
 * T21 (plan §3.2): main branch の 498 行から PR #291 head 2870 行へ 5.6x 肥大化し、
 * 同一 key 重複が多発（`Configuration Library` 4 回 / `Setup step` vs `Setup Step` /
 * `Add Custom Action` vs `Add custom action` 等）していたため、lint gate として導入。
 *
 * Duplicate policy:
 *   (a) Exact duplicate (byte-identical key) — hard error
 *   (b) Case-variant duplicate (same key modulo case) — hard error
 *   (c) Whitespace-normalized duplicate — hard error
 *
 * Usage:
 *   node scripts/check_glossary_duplicates.mjs          # lint mode, exit 2 on duplicate
 *   node scripts/check_glossary_duplicates.mjs --list   # list all duplicates without exit
 *
 * Exit codes:
 *   0 — no duplicates
 *   2 — duplicates found
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import { isDirectRun } from '../lib/cli.mjs';

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../..');
const GLOSSARY_PATH = path.join(ROOT, 'docs/GLOSSARY.md');

function parseGlossaryEntries(markdown) {
  const lines = markdown.split('\n');
  const entries = [];
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    // Skip table header / separator / non-entry rows
    if (!line.startsWith('| ')) continue;
    if (line.startsWith('| --- ') || line.startsWith('| 用語 ')) continue;
    // Format: | term | description |
    const match = line.match(/^\|\s*([^|]+?)\s*\|\s*([^|]+?)\s*\|$/);
    if (!match) continue;
    const term = match[1].trim();
    // Skip obvious header rows
    if (term === '用語' || term === 'term') continue;
    entries.push({ line: i + 1, term, description: match[2].trim() });
  }
  return entries;
}

function normalizeKey(term) {
  return term.toLowerCase().replace(/\s+/g, ' ').trim();
}

function findDuplicates(entries) {
  const byNormalized = new Map();
  for (const entry of entries) {
    const key = normalizeKey(entry.term);
    if (!byNormalized.has(key)) byNormalized.set(key, []);
    byNormalized.get(key).push(entry);
  }
  const duplicates = [];
  for (const [key, group] of byNormalized) {
    if (group.length > 1) {
      duplicates.push({ normalizedKey: key, entries: group });
    }
  }
  return duplicates;
}

function main() {
  const args = process.argv.slice(2);
  const listMode = args.includes('--list');

  const md = fs.readFileSync(GLOSSARY_PATH, 'utf8');
  const entries = parseGlossaryEntries(md);
  const duplicates = findDuplicates(entries);

  if (duplicates.length === 0) {
    console.log(`OK: ${entries.length} entries, no duplicates detected.`);
    process.exit(0);
  }

  console.error(`DUPLICATES: ${duplicates.length} duplicate groups detected in ${GLOSSARY_PATH}`);
  for (const dup of duplicates) {
    console.error(`\n  "${dup.normalizedKey}" (${dup.entries.length} occurrences):`);
    for (const e of dup.entries) {
      console.error(`    L${e.line}: | ${e.term} | ${e.description} |`);
    }
  }
  console.error(
    `\nResolution: merge duplicates into 1 entry (pick the most precise description and delete the rest).`,
  );
  if (listMode) process.exit(0);
  process.exit(2);
}

if (isDirectRun(import.meta.url)) {
  main();
}

export { parseGlossaryEntries, normalizeKey, findDuplicates };

File: scripts/py/src/testim_parity/tools/check_glossary_duplicates.py
"""``scripts/tools/check_glossary_duplicates.mjs`` の Python port。

``docs/GLOSSARY.md`` の重複エントリ (exact / case-variant / whitespace-variant)
を検出する lint gate。

Exit codes:

- 0 — 重複なし
- 2 — 重複あり (``--list`` 指定時は 0)

Duplicate policy (normalize_key で lower + whitespace-collapse してから比較):

- (a) Exact duplicate (byte-identical key) — hard error
- (b) Case-variant duplicate (same key modulo case) — hard error
- (c) Whitespace-normalized duplicate — hard error
"""

from __future__ import annotations

import argparse
import re
import sys
from pathlib import Path

from ..project import ROOT_DIR

__all__ = [
    "find_duplicates",
    "main",
    "normalize_key",
    "parse_glossary_entries",
]


_GLOSSARY_REL = Path("docs/GLOSSARY.md")
_ROW_RE = re.compile(r"^\|\s*([^|]+?)\s*\|\s*([^|]+?)\s*\|$")


def parse_glossary_entries(markdown: str) -> list[dict[str, object]]:
    """glossary markdown から entries を抽出する (mjs 等価)。"""
    lines = markdown.split("\n")
    entries: list[dict[str, object]] = []
    for i, line in enumerate(lines):
        if not line.startswith("| "):
            continue
        if line.startswith("| --- ") or line.startswith("| 用語 "):
            continue
        match = _ROW_RE.match(line)
        if not match:
            continue
        term = match.group(1).strip()
        if term in ("用語", "term"):
            continue
        entries.append({"line": i + 1, "term": term, "description": match.group(2).strip()})
    return entries


def normalize_key(term: str) -> str:
    """lowercase + whitespace collapse (mjs 等価)。"""
    return re.sub(r"\s+", " ", term.lower()).strip()


def find_duplicates(entries: list[dict[str, object]]) -> list[dict[str, object]]:
    """normalize_key 後に同一キーを持つ entry group のみ残す (mjs 等価)。"""
    by_normalized: dict[str, list[dict[str, object]]] = {}
    for entry in entries:
        key = normalize_key(str(entry["term"]))
        by_normalized.setdefault(key, []).append(entry)
    return [
        {"normalizedKey": key, "entries": group}
        for key, group in by_normalized.items()
        if len(group) > 1
    ]


def main(argv: list[str] | None = None) -> int:
    """CLI エントリポイント。exit code (0 or 2) を返す。"""
    parser = argparse.ArgumentParser(description="Detect duplicate entries in docs/GLOSSARY.md")
    parser.add_argument(
        "--list",
        action="store_true",
        help="重複を表示するが exit 0 を返す (CI で fail させない用)",
    )
    args = parser.parse_args(argv)

    glossary_path = ROOT_DIR / _GLOSSARY_REL
    md = glossary_path.read_text(encoding="utf-8")
    entries = parse_glossary_entries(md)
    duplicates = find_duplicates(entries)

    if not duplicates:
        print(f"OK: {len(entries)} entries, no duplicates detected.")
        return 0

    print(
        f"DUPLICATES: {len(duplicates)} duplicate groups detected in {glossary_path}",
        file=sys.stderr,
    )
    for dup in duplicates:
        key = dup["normalizedKey"]
        group = dup["entries"]
        assert isinstance(group, list)
        print(f'\n  "{key}" ({len(group)} occurrences):', file=sys.stderr)
        for e in group:
            print(
                f"    L{e['line']}: | {e['term']} | {e['description']} |",
                file=sys.stderr,
            )
    print(
        "\nResolution: merge duplicates into 1 entry "
        "(pick the most precise description and delete the rest).",
        file=sys.stderr,
    )
    return 0 if args.list else 2


if __name__ == "__main__":
    sys.exit(main())
